// react-query-kit/build/lib/router.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var createInfiniteQuery = require('./createInfiniteQuery.js');
var createMutation = require('./createMutation.js');
var createQuery = require('./createQuery.js');
var createSuspenseInfiniteQuery = require('./createSuspenseInfiniteQuery.js');
var createSuspenseQuery = require('./createSuspenseQuery.js');

const buildRouter = (keys, config) => {
  return Object.entries(config).reduce((acc, [key, opts]) => {
    if (!opts._type) {
      acc[key] = buildRouter([...keys, key], opts);
    } else {
      const options = { ...opts,
        [opts._type === "m" ? "mutationKey" : "queryKey"]: [...keys, key]
      };
      acc[key] = opts._type === "m" ? {
        useMutation: createMutation.createMutation(options),
        ...createMutation.createMutation(options)
      } : opts._type === "q" ? {
        useQuery: createQuery.createQuery(options),
        useSuspenseQuery: createSuspenseQuery.createSuspenseQuery(options),
        ...createQuery.createQuery(options)
      } : {
        useInfiniteQuery: createInfiniteQuery.createInfiniteQuery(options),
        useSuspenseInfiniteQuery: createSuspenseInfiniteQuery.createSuspenseInfiniteQuery(options),
        ...createInfiniteQuery.createInfiniteQuery(options)
      };
    }

    return acc;
  }, {
    getKey: () => keys
  });
};

const router = (key, config) => {
  return buildRouter(Array.isArray(key) ? key : [key], config);
};

function query(options) {
  return { ...options,
    _type: 'q'
  };
}

function infiniteQuery(options) {
  return { ...options,
    _type: 'inf'
  };
}

function mutation(options) {
  return { ...options,
    _type: 'm'
  };
}

router.query = query;
router.infiniteQuery = infiniteQuery;
router.mutation = mutation;

exports.router = router;
